### PR TITLE
[TKW] Mapping dynamic offset API

### PR DESCRIPTION
**Pull request by Hardcode84**
New mapping API and test to show intended usage.

### CODE REVIEW

**Comment by harsh-nod** [comment]
Please rebase on top of master to get the builders working.

File: iree/turbine/kernel/wave/codegen.py [hunk 1]
```diff
@@ -642,12 +657,22 @@ def _construct_gather_scatter_indices(
 
         offsets.append(IntegerAttr.get(IndexType.get(), offset))
 
+    def extract0(src):
+        static_pos = [0 for i in range(src.type.rank)]
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/codegen.py:661]: minor: how about static_pos = [0] * src.type.rank

**Review by harsh-nod — CHANGES_REQUESTED** [review]
Can you add a lit test so we can see what the IR should look like?

File: iree/turbine/kernel/lang/wave_types.py [hunk 2]
```diff
@@ -167,9 +167,17 @@ class IndexMapping:
...
+    dynamic_vals: dict[IndexSymbol, int]
 
     def __init__(
-        self, num_iterators: int, inputs: SymbolsMap, outputs: SymbolsMap
+        self,
+        num_iterators: int,
+        inputs: SymbolsMap,
+        outputs: SymbolsMap,
+        dynamic_val_mappings: SymbolsMap
+        | list[SymbolsMap, ...]
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply iree/turbine/kernel/lang/wave_types.py]: Yeah, good idea

File: iree/turbine/kernel/lang/wave_types.py [hunk 3]
```diff
@@ -191,11 +199,23 @@ def __init__(
         self.iteration_shape = iter_shape
         self.input_mapping = inputs
         self.output_mapping = outputs
+        if dynamic_val_mappings is None:
+            dynamic_val_mappings = ()
+        elif not isinstance(dynamic_val_mappings, (list, tuple)):
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply iree/turbine/kernel/lang/wave_types.py]: done

File: iree/turbine/kernel/lang/wave_types.py [hunk 4]
```diff
@@ -191,11 +199,23 @@ def __init__(
         self.iteration_shape = iter_shape
         self.input_mapping = inputs
         self.output_mapping = outputs
+        if dynamic_val_mappings is None:
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply iree/turbine/kernel/lang/wave_types.py]: good idea

File: iree/turbine/kernel/lang/wave_types.py [hunk 5]
```diff
@@ -191,11 +199,23 @@ def __init__(
...
+            dynamic_val_mappings = ()
+        elif not isinstance(dynamic_val_mappings, (list, tuple)):
+            dynamic_val_mappings = (dynamic_val_mappings,)
+
+        self.dynamic_val_mappings = tuple(dynamic_val_mappings)
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply iree/turbine/kernel/lang/wave_types.py:206]: We can accept list or tuple, but inside the object I want to always store it as tuple.

File: iree/turbine/kernel/lang/wave_types.py [hunk 6]
```diff
@@ -191,11 +199,23 @@ def __init__(
...
+        num_dyn_vals = len(dynamic_val_mappings)
+        self.dynamic_vals = {self.dynamic_val(i): i for i in range(num_dyn_vals)}
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply iree/turbine/kernel/lang/wave_types.py]: done

File: tests/kernel/wave/wave_e2e_test.py [hunk 7]
```diff
@@ -290,6 +292,164 @@ def test(
...
+            mapping=mapping,
+            mapping_dynamic_vals=(offset,),
+            elements_per_thread=ELEMS_PER_THREAD,
+        )
+
+    config = get_default_run_config()
+
+    a = device_randn(shape, dtype=torch.float16)
+    off = (
+        device_randperm(shape[0], dtype=torch.int32)
```
[the remark this answers is not in the document]
Comment (harsh-nod) [reply tests/kernel/wave/wave_e2e_test.py]: makes sense, thanks!

File: iree/turbine/kernel/lang/wave_types.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (harsh-nod) [reply iree/turbine/kernel/lang/wave_types.py]: thanks

File: lit_tests/kernel/wave/codegen.py [hunk 8]
```diff
@@ -383,6 +383,70 @@ def test(
...
+        # CHECK:            %[[D10:.*]] = stream.binding.subspan %[[ARG0]][%[[C0]]] : !stream.binding -> memref<16x16xf16, strided<[16, 1], offset: ?>>
+        # CHECK:            %[[D11:.*]] = arith.index_cast %[[D9]] : vector<16xi32> to vector<16xindex>
+        # CHECK:            %[[D12:.*]] = arith.muli %[[D11]], %[[CST0]] overflow<nsw, nuw> : vector<16xindex>
+        # CHECK:            %[[D13:.*]] = vector.splat %{{.*}} : vector<16xindex>
+        # CHECK:            %[[D14:.*]] = arith.addi %[[D13]], %[[D12]] overflow<nsw, nuw> : vector<16xindex>
+        # CHECK:            %[[D15:.*]] = vector.splat %{{.*}} : vector<16xindex>
+        # CHECK:            %[[D16:.*]] = arith.addi %[[D14]], %[[D15]] overflow<nsw, nuw> : vector<16xindex>
+        # CHECK:            %[[D17:.*]] = arith.addi %[[D16]], %[[CST1]] overflow<nsw, nuw> : vector<16xindex>
+        # CHECK:            %[[D18:.*]] = vector.constant_mask [16] : vector<16xi1>
+        # CHECK:            %[[D19:.*]] = vector.gather %[[D10]][%[[C0]], %[[C0]]] [%[[D17]]], %[[D18]], %[[CST]] : memref<16x16xf16, strided<[16, 1], offset: ?>>, vector<16xindex>, vector<16xi1>, vector<16xf16> into vector<16xf16>
```
Comment (harsh-nod) [inline lit_tests/kernel/wave/codegen.py:445]: thanks, nice to see the gather and how its being used :)

**Review by harsh-nod — APPROVED** [review]
lgtm! thanks!

**Pull request merged**